### ISSUE TITLE
v2 Specification

### ISSUE THREAD

**Issue by neilcampbell** [SEEK-Jobs/pact-net#9]
Having this feature will be very useful for us. We are planning on building the regex matching to conform with the [v2 spec gist](https://gist.github.com/bethesque/5a35a3c1cb9fdab6dce7).

**Comment by kevmee** [comment]
How is this going? Happy to help out.

**Comment by neilcampbell** [comment]
@kevmee I have done a small PoC, however nothing concrete. I am currently working on finalising the v1 release with getting the test output nice to use. That would be great if you want to help out.

**Comment by kevmee** [comment]
Sweet. Send me over what you have. I'll start looking at the source code today. I'm hoping to roll out PactNet here and I believe this feature would be awesome.

**Comment by neilcampbell** [comment]
Nice!

The details can be found on the branch "feature/regex-response-matching". It's pretty stale, and we haven't fleshed out much, we were just trying to ascertain what the DSL part may look like. It's best to pretty much start again.

I would start with rolling out Pact and seeing if you actually need regex support. You can actually get pretty far with just using the provider state support and inserting deterministic test data (on the provider side). The only reason we haven't built regex support is that we actually don't need it at the moment.

Are you able to share where you are rolling it out?

**Comment by neilcampbell** [comment]
Some more discussion about spec v2 and regex matching.

https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/pact-dev/KIL2IjeS7RY/MaGi_t7i_gMJ

**Comment by kevmee** [comment]
I'm working at an insurance company, we're heading down the micro services path so want to get this set up asap.

As you suggested I will be rolling it out, but do see this as a limitation - unless of course I'm missing something that caters for the following scenario:

Consumer -> POST request
Service -> returns 201 with {id=XYZ}

I need to know id exists in the body but have no idea what the value will be. I would prefer not to use a provider to insert an expected id as I would not be testing the *real* API. I would be changing the API behavior when the tests are run, which cheapens the value of the tests.

**Comment by neilcampbell** [comment]
Yeah the id scenario is a good example.

The only reason you wouldn't know what the id was, is if you could not make the data at the time the test is run, deterministic. Having deterministic state makes integrated testing much easier and way less fragile.

We are not changing the behaviour of the API in this instance, we are making the state the API uses deterministic. This is generally a sign of a good test.

**Comment by kevmee** [comment]
I understand the benefit of deterministic state when mocking an external service for the sake of an integration test however I'm unsure about altering the internal state of the service that is actually under test, in this case the API.

Anyway I will be doing some work to roll this out over the next week or so. I hope to take a look at this feature at the end of next week.

**Comment by neilcampbell** [comment]
This is where it gets confusing, the purpose of Pact is to validate the contract between the consumer and provider using a consumer driven approach, not to necessarily provide a testing tool for the provider itself. However as you mention, it is actually testing the provider code to some extent as well. You should still test the provider using normal unit style approaches etc.

The Ruby Pact GitHub wiki is worth checking out as well, as it has more info about these kind of things.
https://github.com/realestate-com-au/pact/wiki/FAQ
https://github.com/realestate-com-au/pact/wiki/Best-practices

@bethesque What are your thoughts on this? Any specific related doco you can point to?

**Comment by neilcampbell** [comment]
@kevmee But yeah, I do agree that regex matching is an awesomely useful feature.

**Comment by bethesque** [comment]
Yes, I agree @neilcampbell, the purpose is to validate the contract, not to be a functional test of the provider, and agree @kevmee that regular expressions are required. An intermediate step is to do "type based" matching, where you can say, this should be a String, this should be an Integer, but I don't care what the value is.

What's the roadmap for types/regexp looking like @neilcampbell?

**Comment by neilcampbell** [comment]
@bethesque Nothing in the works at the moment, however @kevmee has offered to start implementing it, which is awesome!

**Comment by bethesque** [comment]
Yay @kevmee! Welcome to team pact. If you haven't already, please read: https://github.com/bethesque/pact-specification/

You're going to need the v2 spec to work from: https://github.com/bethesque/pact-specification/tree/version-2

Specifically, the regular expression specs: https://github.com/bethesque/pact-specification/blob/version-2/testcases/request/body/matches%20with%20regex.json

Be aware that this spec is still in flux! View this as extra incentive to decouple the implementation from the serialisation format ;)

**Comment by kevmee** [comment]
Thanks @bethesque, for the welcome and the links :)

 @neilcampbell I'm hoping to start looking at this next week, so expect to hear a bit more from me then :) As @bethesque mentioned I will look at type based matching as an intermediate step. 

Cheers

**Comment by neilcampbell** [comment]
@kevmee Awesome stuff!

**Comment by bethesque** [comment]
Here's a type based matching example: https://github.com/bethesque/pact-specification/blob/version-2/testcases/request/body/matches%20with%20type.json

**Comment by bethesque** [comment]
The idea is that the match type specified for a node is supposed to then carry down for all children, however, that example does not show it well.

**Comment by neilcampbell** [comment]
@abhayachauhan Just thought I would include you in this conversation about regex matching as I know you guys were also keen to help out with this.

**Comment by abhayachauhan** [comment]
I've been nagging @neilcampbell about this!
Excellent to see some movement on it - we would really love this feature included - but we've worked around it for the moment.
I will be looking into this in the coming fortnight as we need it fairly soon!

**Comment by bethesque** [comment]
@uglyog and I implemented type based matching for arrays on Friday for the JVM and Ruby impls.

**Comment by kevmee** [comment]
Hey folks, I'll be working on setting up pact here before getting around to this. Will keep you posted when I get a chance to get started.

**Comment by neilcampbell** [comment]
@kevmee :+1:

**Comment by abhayachauhan** [comment]
@neilcampbell @bethesque started working on this (type matching only)

consumer side json updated:
https://github.com/abhayachauhan/pact-net/blob/type-matching/Samples/EventApi/Consumer.Tests/pacts/consumer-event_api.json

have I missed anything? on the right track?

**Comment by bethesque** [comment]
Not quite there. "requestMatchingRules" is now just "matchingRules". `"match" : "true"` isn't a thing, I think you mean `"match" : "type"`? Check this example out. https://github.com/bethesque/pact-specification/blob/version-2/testcases/request/body/matches%20with%20type.json

**Comment by neilcampbell** [comment]
Hey @abhayachauhan!
On top of what @bethesque we want to have a DSL on the consumer for generating the matching expressions. @bethesque has a nice gist with what the ruby DSL may look like, we could do something similar (however we cant use functions in the same way). We did a small PoC in the feature/regex-response-matching branch (it is way behind and very stale, so only use it as a rough guide) https://github.com/SEEK-Jobs/pact-net/blob/feature/regex-response-matching/Samples/EventApi/Consumer.Tests/EventsApiConsumerTests.cs#L248 (highlighted line). You guys may be able to come up with a nicer approach though.

Also I just noticed that your fork is behind by 44 commits, so please patching your changes on top of the current upstream master. I have done a lot of refactoring whilst getting v1 ready, so it is worth syncing before continuing.

Shout if you need anything else!

**Comment by abhayachauhan** [comment]
Thanks @neilcampbell and @bethesque 
I was offline when working so I didn't have your example with me, working from memory.
I've realigned my fork and done some refactoring. I have had a good look at your spike and it seems like a great starting point. I've refactored , I'll touch base as I progress.

**Comment by abhayachauhan** [comment]
@neilcampbell @bethesque I've refactored the consumer side. Neil let me know what you think - it ended up fairly similar to your spike.

Here is the pact - https://gist.github.com/abhayachauhan/68e802898b3d566f8acb

I'm looking at the provider side - I'm starting to see a lot of areas which will need refactoring... especially getting matchers into comparers.

Its all a bit rough at the moment, and will be ironed out once its working - let me know if I've missed anything.

**Comment by neilcampbell** [comment]
Awesome stuff, looking pretty good! I will take a closer look on the weekend at the branch in your fork.

@bethesque With regards to the spec part, should we be following https://gist.github.com/bethesque/5a35a3c1cb9fdab6dce7 or https://github.com/bethesque/pact-specification/blob/version-2/example.json (I just noticed this one)?

Yeah the provider side is likely where the bulk of the work is. The body comparer is particularly ugly and needs a good refactor as you mention.

**Comment by abhayachauhan** [comment]
Finished the regex and type matching support end to end. 
Working well I believe. 

Just need to go through and refactor (mainly the provider side as discussed).

Here is an example of the pact:
https://gist.github.com/abhayachauhan/70c14395e9f1f8a6d819

Let me know what you think @bethesque @neilcampbell 

I've made some assumptions on the type matching side of things. Strings/Ints/Booleans are all straight forward.
For example an expected Object matches an actual Object = pass. Properties are not evaluated at all. 
Same with arrays. We can discuss further if required.

**Comment by bethesque** [comment]
@neilcampbell Neither example. Those were just pseudo code, thinking on a gist. You should be using the pact-specification version 2 tests https://github.com/bethesque/pact-specification/tree/version-2/testcases. For example, the ruby code has a dynamic test that loops through each test case an checks that it agrees with the 'match' element in the fixture. https://github.com/bethesque/pact-support/blob/master/spec/pact_specification/compliance-2.0.rb

I think you already had one of these tests for version 1?

Btw, the node "responseMatchingRules" should be "matchingRules". 

I'm not sure what you mean by "Object matches an actual Object", properties not evaluated @abhayachauhan. I'm don't think I understand what you mean by "properties", maybe it's a .net term. Can you give a JSON example?

**Comment by abhayachauhan** [comment]
Sounds like I've been working off pseudo code :P Back to the drawing board!

@neilcampbell I'm assuming we don't have any data driven tests which run all these test cases seeing we're not v2 compliant. Do we have any data driven tests like these?

Looking through the tests, it seems to have a lot more functionality than I expected. Need to get my head around some of these...
1. Matching Rules for request and response (Response I understand, but Request?)
2. Wildcards in the path (Eg: https://github.com/bethesque/pact-specification/blob/version-2/testcases/request/body/array%20with%20nested%20array%20that%20matches.json) - may need some clarification on how these wildcards work?
3. Min matching for arrays

Any examples of a v2 pact?

**Comment by neilcampbell** [comment]
@abhayachauhan Yep we have the specification data driven tests setup and ready to go. You will just need to update the Git submodule that references the spec repo to point to the version-2 branch on your fork/branch. See https://github.com/SEEK-Jobs/pact-net/tree/master/PactNet.Tests/IntegrationTests/Specification

**Comment by neilcampbell** [comment]
@abhayachauhan And yep there is a fair number of rules to the matching. Took me a while to get it right :) The readme's on the specification repo are worth have a good read through.

**Comment by neilcampbell** [comment]
@bethesque Yep we already have those tests setup. There is a git submodule linked into the pact-net repo.

**Comment by bethesque** [comment]
@abhayachauhan
 
> 1. Matching Rules for request

Flexible request matching is more useful when Pact is used for higher level tests (eg. driven from the UI). If the matching is exact, the tests are brittle, and the number of interactions you end up needing to verify explodes. Have a read of this:  https://github.com/realestate-com-au/pact/wiki/Best-practices#use-pact-for-isolated-unit-tests-think-carefully-about-how-you-use-it-for-integrated-tests

> 2. Wildcards in the path

"*" means "match every element in the array", as distinct from `[n]` which is, match the nth element.

The test cases *are* examples of pacts (or at least, parts of the pact).

**Comment by bethesque** [comment]
Even the ruby version is not completely v2 compliant, but starting with the regular expressions and type based matching will get you a long way.

**Comment by abhayachauhan** [comment]
@neilcampbell I've implemented regex/type matching for responses so far. 
Doesn't handle wildcards yet, nor request matching. 
Done some refactoring, probably needs another look once all the matching functionality is complete.
Could become a very large PR if I try to take the whole thing on.
Thoughts?

**Comment by neilcampbell** [comment]
@abhayachauhan Awesome! Going to hopefully take a look at this today, and will get back to you.

**Comment by neilcampbell** [comment]
@abhayachauhan I had a look tonight and it's starting to look pretty good. I think it's probably worth teeing up a skype chat/code review session/brainstorm some ideas. Thoughts?

**Comment by abhayachauhan** [comment]
Yeah it is worth doing that - there are a lot of rough edges. It would be good to chat about it.

I've also started looking at the wildcard matching, I'm finding the JSONPath interpretation seems a little different - having to manipulate JSON paths which is not fun.

Once I check that in (next day or two), I'll drop you a line.

Min matching isn't too far off either.

**Comment by neilcampbell** [comment]
@abhayachauhan Sounds good, looking forward to the chat.

**Comment by abhayachauhan** [comment]
@neilcampbell Having some trouble coming up with a nice dsl for wildcards.
When are you free to chat? Prefer office hours or after?

**Comment by neilcampbell** [comment]
@abhayachauhan Sorry we have a hackathon on today and tomorrow, so things are a bit crazy. Are you free on Sat for a chat?

**Comment by abhayachauhan** [comment]
I'm usually free 8:30pm onwards - after the kids are asleep. Saturday should be fine.

**Comment by neilcampbell** [comment]
@abhayachauhan Can't do tonight or Sunday night. Let's move this to twitter DMs.

**Comment by abhayachauhan** [comment]
@neilcampbell Another interesting case that I've encountered with matching - 

https://github.com/bethesque/pact-specification/blob/version-2/testcases/response/body/array%20at%20top%20level%20with%20matchers.json

dob field has a regex match - where JsonNet converts to datetime automatically - this will obviously fail. If we switch Datetime parsing off, we lose the ability to do type matching for dates on this.
Probably should turn it off and treat it as a string...

**Comment by neilcampbell** [comment]
@abhayachauhan Yeah I think it should just be treated as a string.

**Comment by neilcampbell** [comment]
Hey @abhayachauhan.
I did a big refactoring of the comparers over the weekend and extracted out the concept of matchers.
So far it is looking pretty good.

I also started working on parsing the interaction body to resolve the matchers, then cleansing the body of the matchers (including nested matcher functions). So far so good and will hopefully complete that tonight or tomorrow.

From there we can start building the different types of matchers (which we can probably extract them from the stuff you have been working on), setup the rules for composing them (with weighting/ specificity), start serialising in the pact file and the consumer side is pretty well be sorted.

**Comment by bethesque** [comment]
Hey guys, don't know if it's just the terms you're using, but you'll find life a whole lot easier down the track if you're working on the concept of a "differ", not a "matcher". The algorithm should take two structures, and return the difference, rather than returning a boolean. This is so that you can then display the differences in a meaningful way to the user. A massive proportion of the usability of the Pact implementations depends on the readability of the diffs.

**Comment by neilcampbell** [comment]
@bethesque Yep it's the terminology we are using. We are actually returning a list of all paths checked in the object graph and if each check was a pass or fail (which will include the diff). That way we can actually override a pass/fail result from a composed matcher if another result is more specific/has a higher weight. Hopefully that makes sense?

**Comment by bethesque** [comment]
Nope, not sure what that means at all! How would a match have a different weight?

**Comment by neilcampbell** [comment]
Maybe I am thinking up things that don't exist :) Take the below example, where we have a type matcher and a regex matcher nested inside the type matcher. This is a pretty bad example, but hopefully I highlights what I am getting at.

```C#
response.Body = new
{
  test = MatcherBuilder.Type(new
  {
    temp = MatcherBuilder.Regex("te", "A regex exp")
  })
}
```

The body returned from the mock provider
```js
{
  "test": {
    "temp": "te"
  }
}
```

If both those matchers are testing something about the $.test.temp path (one via a wildcard on the type matcher and one via the direct path on the regex matcher) and we get conflicting results from each of those matchers (type fails, regex passes for example), is it a pass or failure? Does any failure on that path always result in an overall failure?

**Comment by bethesque** [comment]
In that case, the type match is not performed, only the regexp match is performed.

**Comment by bethesque** [comment]
I'd expect a structure like this to be returned:
```
{
  test: {
    temp: new RegularExpressionDifference(expected: "a regexp exp", actual: "te")
  }
}
```

**Comment by neilcampbell** [comment]
I think this one is easier to talk through, however I think we are both on the same page.

**Comment by abhayachauhan** [comment]
@neilcampbell excellent work dude - lets tee up a chat when your done

**Comment by neilcampbell** [comment]
@abhayachauhan I pushed the changes I have been working on to the version-2 branch https://github.com/SEEK-Jobs/pact-net/tree/version-2. It's still a WIP (so there are some obvious things to cleanup), however would love some feedback with regards to the approach and how it fits with the stuff you are working on.

**Comment by abhayachauhan** [comment]
@neilcampbell awesome..  Shall check it out in in the next day.

**Comment by abhayachauhan** [comment]
@neilcampbell looking really good! :+1: 

Noticed some things -
ProviderServiceResponseComparer should use expected.MatchingRules rather than actual?
TypeMatcher.Type points to Regex Type

Let me know when you'd like help.

**Comment by neilcampbell** [comment]
@abhayachauhan Yep good finds!
You can start patching the matching rules you are working on in and wiring the rest of the bits up. I would start with 1 matching rule and get that working all the way, then add the next etc. Happy to help with any of this stuff. May even be worth another chat.

**Comment by finnnewick-justgiving** [comment]
Hi there - great work on the .net version of pact!

I'm looking at using pact in some projects but lack of regex or type support will make things tricky. How close are things to working? I'm happy to pick some stuff up if that'd help move things closer to the v2 spec.

**Comment by neilcampbell** [comment]
@finnnewick-justgiving Sorry I have taken a little while to respond.
To be honest, I can't see myself getting any free time soon to get this implemented. So yeah it could be a little while off, as I have quite a few other things that need my attention at the moment.

Happy for you to take a look and see what you think? It's definitely something that you could work on, even if it's just the regex piece. I did some work a while back to make this easier (hopefully) to implement.

**Comment by finnnewick-justgiving** [comment]
OK, cool. I can't commit to anything yet but I'll start talking about it with my team and see if we can get something working.

**Comment by neilcampbell** [comment]
Awesome, sounds great! The version-2 branch of the specification project is worth look at to help give you some context about the change. https://github.com/pact-foundation/pact-specification/tree/version-2

**Comment by robertfall** [comment]
Hi Neil,

Sorry it's taken so long to get back to this. We've had to prioritize sprint work over this. We've also found that hard value dependencies in our Provider tests haven't been such a big problem.

Ideally we would like to see type based pacts in our system. Our Scala consumers currently use a DSL to build up type based pacts. Our .NET code currently uses the automatic object serialization in Pact.net. The object serialization is great in some aspects as it provides a 1 to 1 mapping between the pact and our actual request and response objects.

We've seen with the DSL's in Scala that it's too easy to add a field to your response object, and thus a dependency on the server, without updating the pact with the DSL.

Therefore I feel the best approach to take would be something like the Ruby library does:
```ruby
will_respond_with(
    status: 200,
    headers: {"Content-Type" => "application/json"},
    body: {
      Pact.like(
        name: "Mary",
        age: 73
      )
    })

**Comment by neilcampbell** [comment]
@robertfall Sorry I am not 100% sure what you mean? When you mention not a fan of describing the pacts using a DSL, are you talking about using a dynamic object to express the Request/Response body?

`We've seen with the DSL's in Scala that it's too easy to add a field to your response object, and thus a dependency on the server, without updating the pact with the DSL.`

In response to the above, I would imagine that if the added field was critical in processing the response, then the consumer should throw some kind of validation error and the consumer test would break.

**Comment by robertfall** [comment]
@neilcampbell that would indeed by great but it means the developer has to remember to update the consumer test every time they add use a new field on the response.

We've seen the following happen in our codebase:

The consumer code deserializes JSON into the response class:
```scala
case class Response(name: string, isActive: bool)
```

The consumer test body DSL looks as follows:

``` scala
val body = new PactDslJsonBody()
      .`stringType`("name")
        .booleanType("isActive")
      .closeObject()
```

The developer then added a new field to the API `isArchived`. The user updated the response class:
```scala
case class Response(name: string, isActive: bool, isArchived: bool)
```
The developer proceeded to use the Response class and depend on isArchived. The consumer and provider tests were never updated and never failed. 

The consumer test deserialized the json generated by the DSL statement (which didn't contain the `isArchived` field) into an instance of the response class leaving the isArchived field empty. The provider test didn't fail as all that happened was a new field was added to the provider which doesn't break a contract. I think we can probably explicitly mark all the fields on the response objects as required in some manner which would prevent this from happening but it hasn't been done in the code base so far. In effect this makes the contracts useless for determining what fields are actually required.

In our .NET codebase however, when a developer adds a field to the response class the consumer test is automatically updated as the mock provider serializes an instance of the response class. I have just realised though that if the mock response in the .NET consumer test doesn't set fields of certain types (Objects for example) that they too won't serialize correctly.

In the end I guess it comes down to discipline when changing request and response objects, thought I do personally feel that generating pact JSON using real request and response objects is more robust and accurate than relying purely on a DSL. 

A `Pact.Like(response);` wrapper seems to be the best of both worlds and is something that exists in the [ruby version](https://github.com/realestate-com-au/pact/wiki/Regular-expressions-and-type-matching-with-Pact#type-matching).

Sorry for rambling here. I do still intend to spend some time on adding type-based matchers. I've had to push that back during our current sprint though.

**Comment by neilcampbell** [comment]
@robertfall Thanks heaps for the additional detail!

Ok, so my personal opinion is that we should all get into the mindset of changing the test first, then updating the consumer to match. I can however understand your point of view.

In the .NET version you can change the JSON serializer settings when calling `PactBuilder.MockService()`. So you should be able to force the default values to be serialised into the Pact, should they be omitted in consumer test.

In regards to `Pact.Like(response)`, I do like the idea of that one, but I do think we should have support for both use cases. I personally like to use dynamic objects.

The tricky part is going to be composing all the different rules to make the matching work like you want.

**Comment by VanessaDa** [comment]
Hi Guys,

I have a JSON pact file with href links that is part of the response body from the consumer. Which looks something like this:

"links": [
              {
                "rel": [
                  "self",
                  "paymentDetail"
                ],
               "href": "/api/uk/account/12345/payment-detail"**
              }
            ]
          },


However the provider  responds with the full resource as below:

"href": "http://localhost:12345/api/uk/customeraccount/123456/payment-detail"

Therefore when I run the provider test and it compares the actual to expected it fails. I am using Pact. Net version 1.1.0 which does not support partial matching as of yet. Is there a work around that anyone knows off that could assist me in which the consumer can ignore the server name?

Regards,
Newbiee

**Comment by scottcowan** [comment]
How's it going with this? I'm new to Pact and it seems like this is needed to test services where you can't define the provider state. I'm willing to help

**Comment by neilcampbell** [comment]
Hi @scottcowan!
Thanks for getting in touch. At the moment there is still nothing happening on v2/3 flexible matching.

I'm interested in hearing more about your specific use case and why provider states cannot be defined?

**Comment by scottcowan** [comment]
Hi Neil,

I've used pact net with the pact broker so 
- on build the pact is published
- on deploy of a consumer the pact is tagged per environment
- on deploy of the provider the pacts a gathered from the broker by the tag
- pact-provider-verifier runs each

- I did skip over provider states as we're using this throughout our environments
- I was looking at type matching or regex to identify which fields a consumer is using without explicitly checking the content of the response.
- I could use provider states but I'm using this throughout the deployment pipeline (which may be wrong) and I can rely on the Ids of requests to be valid over time.

**Comment by neilcampbell** [comment]
Usually Pact verification is done before deployment as it let's you know if the consumer/provider pair are compatible and therefore prevents you from deploying an incompatible configuration. The trick to successful provider verification is being able to control the application state (especially with v1 of Pact).

With the request id example you mention, is that an important piece of the contract from a consumer PoV? If it's not you can completely omit that from the Pact and the verification process will just ignore that field.

**Comment by baguaJerry** [comment]
Scott,
Since we currently are using pact for messaging and not http calls we are looking into just implementing that support right now.

Jerry

**Comment by neilcampbell** [comment]
@baguaJerry Cool, i'm interested in hearing more.

**Comment by iamdavidfrancis** [comment]
Hi all,
What is the current state of the v2 progress? I'm interested in using it and am willing to help contribute.

**Comment by neilcampbell** [comment]
Hi @iamdavidfrancis 
Thanks for getting in touch.

Nothing has really changed on this front yet. I am going to sit down on the weekend and work out a path forward. Hopefully I will update this thread on the weekend with some more details on how you can help out 😄

**Comment by bethesque** [comment]
Yay @iamdavidfrancis!

**Comment by cheesemacfly** [comment]
Really interested in having a v2 implementation in .net too. I can help if needed!

**Comment by lopes22** [comment]
I would also like to help. I'm particularly interested in at least getting the provider verification implemented for version 2. We have another team generating pact interactions through the Swift implementation and would like to verify them on our .NET web service.

Has anyone thought of a strategy for handling the "matchingRules" part of the spec? We need to apply the matching rules, but default back to simple equality comparisons for all other paths not included. Json.NET allows us to select tokens given a path expression but there is no way to check if a path would be matched by a path expression. I was thinking maybe we apply all matching rules, keeping track of which paths were matched. Then traverse the Json, checking if the current path was already checked from a matchingRule and if it has been checked already, skip the equality comparison.

**Comment by cnewham** [comment]
@baguaJerry and I put a good bit of work to support the V2 matchingRules and messaging for V3, but have not fully added support for V2 in the current codebase. It's pretty close, but still needs some work to integrate the new matching logic into interactions. I'm hoping to jump back onto this at some point, but if you're interested please have a look at #79

**Comment by neilcampbell** [comment]
I have been thinking this through a bit and chatting with with some of the others in the pact foundation group. Our goal is to eventually have a language agnostic "core" and then simpler language bindings into the core. This will allow us to add new features across the board at lot faster and is why the Rust https://github.com/pact-foundation/pact-reference implementation is being written.

With this in mind, I am playing with idea of doing something similar to what the Pact JS project does and wrap the Ruby implementation. This will give us v2 matching, align the .NET project more towards that goal, and also should make adding .NET core support easier.

What does everyone think of this idea?

**Comment by abhayachauhan** [comment]
I think it is a great idea. One of the challenges moving forward was always going to be keeping all these independent language libraries aligned and moving forward. The tests provided some sort of sanity but the feature parity was always an issue.

The fastest and future proof way to keep aligned. 
Although I'm not sure why Rust was selected; regardless, +1 from me.

**Comment by neilcampbell** [comment]
@abhayachauhan Rust was a good candidate because it works cross platform and has a very lightweight runtime (like C, C++) and can be embedded and called easily from other languages using FFI.

**Comment by gruntster** [comment]
Aye, I think the premise definitely has merit.  I looked at a similar approach last year but with the JVM version.  I tried a couple of things like using the JVM Pact Server with .NET and even converting the Java code to C# using tools like IKVM and Sharpen.  It certainly wasn't trivial.  Perhaps the Rust version will play out a bit nicer.

**Comment by AnnejanBarelds** [comment]
I like the idea. I was/am considering to contribute to implementing new spec versions, so anything that can help speed things up in this regard is great news as far as I'm concerned.

**Comment by iamdavidfrancis** [comment]
I like this idea of wrapping a "core" library for other languages. If we can build out a dll from the Rust implementation, we can extern it easily in C# and have feature parity.

**Comment by cwilcox-fl** [comment]
For those interested in this idea but unfamiliar with FFI/embedding/wrapping (like me):
FFI stands for **Foreign Function Interface**
A good place to start an exploration of this space might be here [Stack Overflow Understanding FFI](https://stackoverflow.com/questions/5440968/understand-foreign-function-interface-ffi-and-language-binding)
If this can work, It's hard for me to express how great an idea I think this is.

**Comment by bethesque** [comment]
If you use the same technique as the javascript library has, you don't even need to use FFI. You just need a .net HTTP client for the mock service, and a nice way to run some provider setup hooks with the pact-provider-verifier on the other end.

**Comment by neilcampbell** [comment]
Yeah totally @bethesque, that is where I am heading to start with. But in my head the FFI piece is in relation to using the Rust integration with other protocols and messaging patterns (V3 spec). We need to get v2 first 😄

**Comment by neilcampbell** [comment]
Also hoping to have a PoC done by the end of the weekend, so fingers crossed!

**Comment by bethesque** [comment]
Awesome.

**Comment by ma499** [comment]
@neilcampbell how did the PoC go? Can contributors help in some way?

**Comment by neilcampbell** [comment]
@ma499 PoC went well. It all works pretty well with a few small trade offs. There will be some breaking changes on the verification side (nothing we can't fix with some doco).

Currently cleaning up a few things, then will push the branch (hopefully in the next couple of days). From there I will compile a list of things to do, and hopefully some of you can help out with them/test the changes in your projects etc.

Then we can release v2 :)

**Comment by neilcampbell** [comment]
I have updated the version-2 branch (https://github.com/SEEK-Jobs/pact-net/tree/version-2) with a new working version that uses the ruby standalone core engine. Inside the repo there is a TODO.txt file which has most of things I think we need to consider before shipping the official version 2 package.

If you are interested in helping out take a look at the list, pick an item, and let me know which one you are working on in this thread.

It would also be great if some people could test the current package and let me know if it all works ok for them. The current nuget build artefact is available at https://ci.appveyor.com/project/SEEKJobs/pact-net/build/175/artifacts

**Comment by bikashbehuria** [comment]
@neilcampbell  Thanks ! I will try to use the MatchingRules as soon as possible. Do you have some sample to look from Consumer and Provider perspective. Or any document to follow for the same ?

**Comment by neilcampbell** [comment]
@bikashbehuria The sample in the repo have been updated, so that's the best place to start.

**Comment by bikashbehuria** [comment]
@neilcampbell Thanks again. I can see the sample.

**Comment by neilcampbell** [comment]
The version-2 branch has now been completed and merged into master. I have also created a beta nuget package which can be installed from nuget.org by running `Install-Package PactNet -Version 2.0.0-beta -Pre`.

Before officially releasing this, I would love to get some people to install/test and make sure you are happy/provide feedback. Let me know if you can help out!

**Comment by neilcampbell** [comment]
I have made an additional tweak so latest is now 2.0.1-beta, which can be installed using `Install-Package PactNet -Version 2.0.1-beta -Pre`

**Comment by bikashbehuria** [comment]
@neilcampbell . Thanks !
I have a question on how we are validating the response in provider side. In earlier version or (may be version 1) , we had ProviderServiceResponseComparer class which is being used for comparing two  i.e. actual and expected. Now there is no such class ? just curious to know , how are comparing now ?  

Just FYI , i am in process to start V2 in my REST Provider and Consumer so thought of asking this quick question.

**Comment by neilcampbell** [comment]
@bikashbehuria The ProviderServiceResponseComparer is an internal class in v1 so I am not 100% sure what you mean? Is there anything you can share in a gist which demonstrates this?

Basically the primary method was to pass a HttpClient, which talked to the provider. Unfortunately that had to be deprecated due to using the ruby core and additionally wanting to get .NET standard/core support.

Instead of passing a HttpClient, now you pass a HTTP URI which can be used to talk with the provider.

**Comment by bikashbehuria** [comment]
@neilcampbell . I was debugging v1 earlier and I could see below code , which was meant for to verify Actual vs Expected   **_public ComparisonResult Compare(ProviderServiceResponse expected, ProviderServiceResponse actual_)**

Where it compares the result coming from actual API call with the JSON PACT file produced by consumer. 

So thought of debugging same and see in latest code V2  as I have some complex data model in my Http JSON response, which I want to verify each attributes. So when I run the current sample in Provider Test Method - EnsureEventApiHonoursPactWithConsumer()  , I am not been able to locate where we are comparing Actual (from API) vs Expected (from PACT file). Also I want to apply some matching rule as I have different consumers , using same API  but they have different matching type.

**Comment by neilcampbell** [comment]
@bikashbehuria Thanks for providing a bit more context, I appreciate you sharing your experience!

Pact-Net is now making an interprocess call into the Ruby standalone shared core to do the actual verification, so you will not see a .NET stack trace for that part.

If you have different consumers calling the same API, they should be seperate Pacts, and as such they should have seperate consumer tests.

**Comment by bikashbehuria** [comment]
@neilcampbell Thanks for all the information. I will setup the tests and come back to you again for any question.

**Comment by cheesemacfly** [comment]
@neilcampbell thanks for the work on the v2 it's great to see this project moving forward! I have been playing a little bit with the beta version and I have 1 question and 1 issue.

It looks like the provider state can now only be a uri? It does require the setup of the middleware like in the readme example but could we use just a string like in v1?

I get an issue about encoding somewhere in the ruby code about my json pact file even though it was working fine with v1. Anything in particular that needs to be done to make it 'compatible' with v2?
(/bin/Debug/pact/lib/vendor/ruby/2.2.0/gems/json-1.8.6/lib/json/pure/parser.rb:242:in `rescue in parse_string': Caught Encoding::CompatibilityError at '∩╗┐{ (JSON::ParserError))

**Comment by bethesque** [comment]
Are you using any unusual characters @cheesemacfly? It looks like '∩╗┐' is the thing that's making it blow up.

When you have a moment @neilcampbell could you set up an end to end example like https://github.com/pact-foundation/pact-ruby-e2e-example that can run on appveyor like https://github.com/pact-foundation/pact-ruby-standalone-windows-test ? It makes it really quick for people to give you an executable example of their problem.

**Comment by bethesque** [comment]
@cheesemacfly could you please raise a separate issue, and include if possible an executable example, and at the very least, the steps to reproduce your issue.

**Comment by neilcampbell** [comment]
@cheesemacfly To answer the question about about just using a string like in v1, unfortunately we cannot do that easily using the Ruby standalone core and also trying to target .NET standard.

**Comment by cheesemacfly** [comment]
@bethesque I actually got a more detailed error message that says: 

> incompatible encoding regexp match (ASCII-8BIT regexp with IBM437 string)

I'll create a seperate ticket with the file I'm using if you think that's worth it.

@neilcampbell Your answer actually triggered a new question: since the target is now .NET standard, I'm assuming the goal is to let people use .NET core and there own choice of os. How is the Ruby standalone going to play on unix systems? I have to admit I'm not a big fan of having such a big dependency on Ruby which reminds me of the mess I had in the past using sass...but I understand the choice was discussed in the messages above. (also I don't really like ruby so I'm definitely biased)

**Comment by neilcampbell** [comment]
@cheesemacfly Yes that is correct. Ruby and travelling ruby (which is what we use for the standalone core) is cross platform. We have builds for all platforms, see https://github.com/pact-foundation/pact-ruby-standalone/releases.

The reason we are doing this is so we can have a common core across all language implementations. At the moment it's ruby, but we would like to potentially move to the rust core, which is much lighter weight. I'm hoping that for the most part the Ruby'isms will stay fairly well invisible.

Having a common core I think it a pretty big deal as it allows us to move quicker and add new features to all languages a lot easier.

**Comment by bikashbehuria** [comment]
@neilcampbell . Not sure , which forum to discuss about one issue I found  but I need some help in starting up my PACT test in consumer side. I am facing some issue in my test , while starting Mock service . Here are some detail , which may help. 
-  Created a Test project using xUnit . I am following what is being created in Event API consumer sample and test.
-  The MockService port is being set as 9222
-  When I am executing the Event API sample , the Mock service is starting up but when I am using my test , its **not** starting up .
- So i thought to debug PactNet with my test project and tried to compare with Event API sample project. Particularly on MockService start part.
- I found that both are  sending same command to start the Ruby core process . However not able to find out why the first one below is failing to start.
 
-p 9222 -l "../../logs/distribution_api_mock_service.log" --pact-dir "../../pacts/" --pact-specification-version "2.0.0" 

-p 9222 -l "../../logs/event_api_mock_service.log" --pact-dir "../../pacts/" --pact-specification-version "2.0.0"

I may be doing some silly mistake but any hint from you would be a great help

Thanks-Bikash

**Comment by bikashbehuria** [comment]
@bethesque . Thanks for your response.
They both ran in different time. I found the reason. Newtonsoft.Json library was not part of my test project . So I added the reference manually , it worked now. 

Thanks !

**Comment by bikashbehuria** [comment]
Just a information to share.
PactNet which is build targeting framework 4  not compatible with Newtonsoft.Json 10.0.* .  I am using latest 10.0.* so I had to change the target build of PactNet to 4.5.2.

**Comment by neilcampbell** [comment]
@bikashbehuria The latest version of PactNet (still beta) does not target .NET 4 any more, but requires min 4.5, see https://www.nuget.org/packages/PactNet/2.0.7-beta.

**Comment by bikashbehuria** [comment]
@neilcampbell  . Thanks . Now I am using latest from Nuget

**Comment by neilcampbell** [comment]
The first official v2 spec release has been created (2.0.17).

To install use either:
- PactNet.Windows (was previously PactNet-Windows)
- PactNet.OSX (was previously PactNet-OSX)
- PactNet.Linux.x86 (was previously PactNet-Linux-x86)
- PactNet.Linux.x64 (was previously PactNet-Linux-x64)
